File: two-pointers/922/922.py
# [AUTHOR]: Peregrin
# [DESCRIPTION]: Python one-pass, O(1) memory, simple code, beats 90%

class Solution:
    def sortArrayByParityII(self, a):
        i = 0 # pointer for even misplaced
        j = 1 # pointer for odd misplaced
        sz = len(a)
        
        # invariant: for every misplaced odd there is misplaced even
        # since there is just enough space for odds and evens

        while i < sz and j < sz:
            if a[i] % 2 == 0:
                i += 2
            elif a[j] % 2 == 1:
                j += 2
            else:
                # a[i] % 2 == 1 AND a[j] % 2 == 0
                a[i],a[j] = a[j],a[i]
                i += 2
                j += 2

        return a

# ===================================================

# [AUTHOR]: votrubac
# [DESCRIPTION]: C++ 5 lines, two pointers + 2-liner bonus

"""

Use two pointers to search for missplaced odd and even elements, and swap them.

vector<int> sortArrayByParityII(vector<int>& A) {
    for (int i = 0, j = 1; i < A.size(); i += 2, j += 2) {
        while (i < A.size() && A[i] % 2 == 0) i += 2;
        while (j < A.size() && A[j] % 2 == 1) j += 2;
        if (i < A.size()) swap(A[i], A[j]);
    }
    return A;
}

Now, some fun for for my minimalistic functional friends. It's techically a two-liner, though I split swap into 3 lines for readability :) It actually may even look a bit cleaner, as you do not have to do "plus 2".

vector<int> sortArrayByParityII(vector<int>& A) {
  for (int i = 0, j = 0; i < A.size() && j < A.size(); ) swap(
      *find_if(begin(A) + i, end(A), [&] (int v) { return (i++ % 2 == 0 && v % 2 != 0) || i == A.size(); }),
      *find_if(begin(A) + j, end(A), [&] (int v) { return (j++ % 2 != 0 && v % 2 == 0) || j == A.size(); }));
  return A;
}

"""

# ===================================================

# [AUTHOR]: haleyysz
# [DESCRIPTION]: [javascript] O(n)

"""

/**
 * @param {number[]} A
 * @return {number[]}
 */
var sortArrayByParityII = function(A) {
    let result = new Array(A.length);
    
    for(let i = 0, even = 0, odd = 1; i < A.length; i ++) {
        if(A[i] % 2 === 0) {
            result[even] = A[i];
            even += 2;
        } else {
            result[odd] = A[i];
            odd += 2;
        }
    }
    return result;
};

"""

# ===================================================


# [AUTHOR] ME
# [DESCRIPTION] I created a shame :) 

class Solution:
    def sortArrayByParityII(self, nums: List[int]) -> List[int]:
        leftPtr = 0
        rightPtr = len(nums) - 1
        
        while leftPtr < rightPtr:
            if self.canMove(nums, leftPtr):
                leftPtr += 1
        
            if self.canMove(nums, rightPtr):
                rightPtr -= 1
                
            if (self.isEven(nums[leftPtr]) and not self.isEven(rightPtr)) and (self.isEven(nums[rightPtr]) and not self.isEven(leftPtr)):
                self.swap(nums, leftPtr, rightPtr)

                leftPtr += 1
                rightPtr -= 1

            
        return nums
            
            
            
    def isEven(self, num):
        return num % 2 == 0
    
    def canMove(self, arr, ptr):
        return self.isEven(ptr) and self.isEven(arr[ptr])
    
    def swap(self, arr, ptr1, ptr2):
        print('> ', arr[ptr1], ptr1, '\n', '>', arr[ptr2], ptr2)

File: two-pointers/922/922.js
// [AUTHOR]: uzzwal
// [DESCRIPTION]: One Pass - Two Pointer Solution

/*

Consider two pointers, one oddIx, pointing to odd position and the other evenIx, pointing to even position.
Begining with oddIx at 1 and evenIx at 0, we will increment these until one exceeds the length of the array.
If we have odd number at even position or even number at odd position, we will like to change this, so we will swap only when both, oddIx and evenIx, holds number of opposite category(even/odd).
Else, we would simply skip the position for oddIx and evenIx if they hold number belonging to their category.


class Solution {
    public int[] sortArrayByParityII(int[] nums) {
        int oddIx = 1, evenIx = 0;
        while(oddIx < nums.length && evenIx < nums.length){
            if (nums[oddIx] % 2 < nums[evenIx] % 2){
                int temp = nums[oddIx];
                nums[oddIx] = nums[evenIx];
                nums[evenIx] = temp;
            }
            if (nums[oddIx] % 2 == 1)
                oddIx = oddIx + 2;
            if (nums[evenIx] % 2 == 0)
                evenIx = evenIx + 2;
        }
        return nums;
    }
}

*/





// ===================================================

// [AUTHOR]: Saura_v
// [DESCRIPTION]: Python 2 pointer solution

/*

# Time complexity - O(n)
# Space complexity - O(1)

class Solution:
    def sortArrayByParityII(self, nums: List[int]) -> List[int]:
        even_idx = 0
        odd_idx = len(nums)-1
        n = len(nums)
        
        while even_idx < n and odd_idx > 0:
            if nums[even_idx] % 2 == 0:
                even_idx += 2
            
            elif nums[odd_idx] % 2 != 0:
                odd_idx -= 2
                
            else:
                nums[even_idx], nums[odd_idx] = nums[odd_idx], nums[even_idx]
                even_idx += 2
                odd_idx -= 2
                
        return nums

*/

// ===================================================

// [AUTHOR]: Edgar__Li
// [DESCRIPTION]: 3 lines JavaScript solution.

let evenArray = A.filter(x => x % 2 === 0);
let oddArray = A.filter(x => x % 2 === 1);
return A.map((x, index) => index % 2 === 0 ? evenArray.pop() : oddArray.pop());


// ===================================================

// [AUTHOR]: jamiehsmith
// [DESCRIPTION]: Javascript - Simple using filter

var sortArrayByParityII = function(A) {
    let evenNums = A.filter((i) => i % 2 === 0);
    let oddNums = A.filter((i) => i % 2 !== 0);
    
    let res = [];
    for (let i = 0; i < A.length / 2; i++) {
        res.push(evenNums[i], oddNums[i]);
    }
    
    return res;
};



// ===================================================


// [AUTHOR] ME
// [DESCRIPTION] The solutions was inspired by uzzwal example
var sortArrayByParityII = function(nums) {
  let oddIdx = 1,
      evenIdx = 0,
      n = nums.length;
  
    
  while (oddIdx < n && evenIdx < n) {
    if (nums[oddIdx] % 2 < nums[evenIdx] % 2) {
      [nums[oddIdx], nums[evenIdx]] = [nums[evenIdx], nums[oddIdx]];
    }
    
    if (isOdd(nums[oddIdx])) {
      oddIdx += 2;
    }
    
    if (isEven(nums[evenIdx])) {
      evenIdx += 2;
    }
  }
    

  
  return nums;
    
};

function isEven(num) {
  return num % 2 === 0;
};

function isOdd(num) {
  return num % 2 !== 0;
};

